Objects/interpolationobject.c

```diff
@@ -0,0 +1,217 @@
+/* t-string Interpolation object implementation */
+
+#include "Python.h"
+#include "pycore_initconfig.h"      // _PyStatus_OK
+#include "pycore_interpolation.h"
+
+static int
+_conversion_converter(PyObject *arg, PyObject **conversion)
+{
+    if (arg == Py_None) {
+        return 1;
+    }
+
+    if (!PyUnicode_Check(arg)) {
+        PyErr_Format(PyExc_TypeError,
+            "%.200s() %.200s must be %.50s, not %.50s",
+            "Interpolation", "argument 'conversion'", "str",
+            arg == Py_None ? "None" : Py_TYPE(arg)->tp_name);
+        return 0;
+    }
+
+    Py_ssize_t len;
+    const char *conv_str = PyUnicode_AsUTF8AndSize(arg, &len);
+    if (len != 1 || !(conv_str[0] == 'a' || conv_str[0] == 'r' || conv_str[0] == 's')) {
+        PyErr_SetString(PyExc_ValueError,
+            "Interpolation() argument 'conversion' must be one of 's', 'a' or 'r'");
+        return 0;
+    }
+
+    *conversion = arg;
+    return 1;
+}
+
+#include "clinic/interpolationobject.c.h"
+
+/*[clinic input]
+class Interpolation "interpolationobject *" "&_PyInterpolation_Type"
+[clinic start generated code]*/
+/*[clinic end generated code: output=da39a3ee5e6b4b0d input=161c64a16f9c4544]*/
+
+typedef struct {
+    PyObject_HEAD
+    PyObject *value;
+    PyObject *expression;
+    PyObject *conversion;
+    PyObject *format_spec;
+} interpolationobject;
+
+/*[clinic input]
+@classmethod
+Interpolation.__new__ as interpolation_new
+
+    value: object
+    expression: object(subclass_of='&PyUnicode_Type')
+    conversion: object(converter='_conversion_converter') = None
+    format_spec: object(subclass_of='&PyUnicode_Type', c_default='&_Py_STR(empty)') = ""
+[clinic start generated code]*/
+
+static PyObject *
+interpolation_new_impl(PyTypeObject *type, PyObject *value,
+                       PyObject *expression, PyObject *conversion,
+                       PyObject *format_spec)
+/*[clinic end generated code: output=6488e288765bc1a9 input=d91711024068528c]*/
+{
+    interpolationobject *self = (interpolationobject *) type->tp_alloc(type, 0);
+    if (!self) {
+        return NULL;
+    }
+
+    self->value = Py_NewRef(value);
+    self->expression = Py_NewRef(expression);
+    self->conversion = Py_NewRef(conversion);
+    self->format_spec = Py_NewRef(format_spec);
+    return (PyObject *) self;
+}
+
+static void
+interpolation_dealloc(interpolationobject *self)
+{
+    PyObject_GC_UnTrack(self);
+    Py_CLEAR(self->value);
+    Py_CLEAR(self->expression);
+    Py_CLEAR(self->conversion);
+    Py_CLEAR(self->format_spec);
+    Py_TYPE(self)->tp_free(self);
+}
+
+static int
+interpolation_traverse(interpolationobject *self, visitproc visit, void *arg)
```

Comment:
```suggestion
interpolation_traverse(PyObject *self, visitproc visit, void *arg)
```

This is an undefined behavior otherwise. So the cast must be done inside the function.

Reply:
Done in [`5782897` (#132662)](https://github.com/python/cpython/pull/132662/commits/5782897688200183b052339bba63767bafe161a0)